contracts/p1/BackingManager.sol

```diff
@@ -146,14 +130,106 @@ contract BackingManagerP1 is TradingP1, IBackingManager {
                     if (req.sellAmount > bal) stRSR.seizeRSR(req.sellAmount - bal);
                 }
 
-                tryTrade(req);
+                openTrade(req);
             } else {
                 // Haircut time
                 compromiseBasketsNeeded(basketsHeld.bottom, basketsNeeded);
             }
         }
     }
 
+    /// Maintain the overall backing policy in an atomic swap via a dutch auction
+    /// @dev Caller must have granted tokenIn allowances for required tokenIn bal
+    /// @param tokenIn The ERC20 token provided by the caller
+    /// @param tokenOut The ERC20 token being purchased by the caller
+    /// @param amountOut {qTokenOut} The exact quantity of tokenOut being purchased
+    /// @return s The exact Swap performed
+    /// @custom:interaction RCEI
+    function swap(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        uint256 amountOut
+    ) external returns (Swap memory s) {
+        // == Refresh ==
+        assetRegistry.refresh();
+        furnace.melt();
+
+        requireCanTrade(0);
+
+        // executeSwap if storage auction already exists
+        if (tradeEnd > block.timestamp) {
+            return executeSwap(dutchAuctions[tradeEnd], amountOut);
+        }
+
+        // === Checks/Effects ===
+
+        require(tradeEnd + dutchAuctionLength > block.timestamp, "no dutch auction ongoing");
+        tradeEnd += dutchAuctionLength;
+
+        // Same TradeRequest from manageTokens()
+        (bool doTrade, TradeRequest memory req) = RecollateralizationLibP1
+            .prepareRecollateralizationTrade(this, basketHandler.basketsHeldBy(address(this)));
+
+        require(doTrade, "swap not available");
+        require(tokenIn == req.buy.erc20() && tokenOut == req.sell.erc20(), "ERC20 swap expired");
+
+        dutchAuctions[tradeEnd] = DutchAuctionLib.makeAuction(
+            req.sell,
+            req.buy,
+            shiftl_toFix(req.sellAmount, -int8(req.sell.erc20Decimals())),
+            minTradeVolume,
+            maxTradeSlippage
+        );
+
+        // === Interactions ===
+
+        // Seize RSR if needed
+        if (req.sell.erc20() == rsr) {
+            uint256 bal = rsr.balanceOf(address(this));
+            if (req.sellAmount > bal) stRSR.seizeRSR(req.sellAmount - bal);
+        }
+        s = executeSwap(dutchAuctions[tradeEnd], amountOut);
+    }
+
+    /// @return The ongoing auction as a Swap
+    function getDutchAuctionQuote() external view returns (Swap memory) {
+        requireCanTrade(0);
+
+        if (tradeEnd > block.timestamp) {
+            return dutchAuctions[tradeEnd].toSwap(progression());
+        }
+
+        require(tradeEnd + dutchAuctionLength > block.timestamp, "no dutch auction ongoing");
```

Comment:
do we want to revert on view functions, or is it better to just return an empty swap? or we can return a swap w/ a bool like we do for `prepareRecollateralizationTrade`?

Reply:
Yeah...good question, it was something I thought about. 

I think I have a very small preference for returning an empty swap, but the implementation trouble I ran into is that this made it harder to re-use logic or checks that otherwise requires in the `swap()` flow. 

If we weren't running up against BackingManager contract size constraints I would say it's worth it, but as-is it would require breaking up `requireCanTrade` more than I would like to.